packages/specs/source/01-app-definition.md

```diff
@@ -1,40 +1,169 @@
+
+```solidity
+struct ETHPaymentChannelState {
+  address alice;
+  address bob;
+  uint256 aBal;
+  uint256 bBal;
+}
+```
+
+## Progressing State
+
+As it has been mentioned before, some kinds of applications require there be a way of progressing some state to a "terminal" state through a series of allowed actions. In these cases, we adopt the model of a state machine that is described via logical states and allowed actions which act as edges between logical states; a "terminal" state is simply one from which there does not exist any outgoing edge (i.e., an "allowed action").
+
+In Counterfactual, if an action wishes to allow its state to be unilaterally progressable, we require the definition of a function that **applies an action to a state to produce a new state** _in addition to_ **a function that determines if an action can be legally taken by a particular turn taker**. As you will see in the [adjudication layer](./02-adjudication-layer.md) section of these specifications, these functions are important in handling on-chain challenge scenarios.
+
+The ultimate purpose of these functions is to ensure the following:
+
+- It should always be extremely explicit what the exact rules of the state channel application that all parties are abiding by are
+- There should always be a single logical turn taker for any given state ([concurrent state updates](#concurrent-state-updates) are disallowed)
+- It should be possible, in the strictly least number of on-chain transactions, use the adjudication layer to fairly resolve a state channel application (without unnecessary gas expenditure)
+
+Here is a helpful diagram for visualizing the nature of such an `applyAction` function:
 
 ![](img/applyAction.svg)
 
-### resolve
+## Resolving State
+
+For any given state channel application is the notion of a resolution. This is a critical concept usually because the resolution can be tied to interesting economic parameters that create the incentives for the behaviour of users in the application to begin with. For example, users will remain online and play a game of Tic-Tac-Toe because they know the rules of the game are deciding who will take home some financial reward. In the Counterfactual framework, this reward is defined in terms of an internal transaction that is executed by the multisignature wallet that is the holder of the state deposits.
```

Comment:
```suggestion
A state channel application defines a resolution. This is a critical concept usually because the resolution can be tied to interesting economic parameters that create the incentives for the behaviour of users in the application to begin with. For example, users will remain online and play a game of Tic-Tac-Toe because they know the rules of the game are deciding who will take home some financial reward. In the Counterfactual framework, this reward is defined in terms of an internal transaction that is executed by the multisignature wallet that is the holder of the state deposits.
```

Reply:
ambiguous modifier "usually"